### PR TITLE
Add and deal with hlint

### PR DESCRIPTION
**Pull request by chiroptical**

### CODE REVIEW

**Review by parsonsmatt — APPROVED** [review]
Looks great!

Do we want to copy a `.hlint.yaml` from our private repo, or do we want to just use the stock one?

File: Makefile [hunk 1]
```diff
@@ -22,4 +22,7 @@ ghcid-test: hpack
 format:
 	find src/ test/ -name "*.hs" -exec ormolu -i {} \;
 
-.PROXY: build test hpack repl repl-test ghcid ghcid-test format
+hlint:
+	hlint .
+
+.PROXY: build test hpack repl repl-test ghcid ghcid-test format hlint
```
[the remark this answers is not in the document]
Comment (chiroptical) [reply Makefile:28]: :rofl: yes!

**Comment by chiroptical** [comment]
Since this is a public repo, I think it makes sense to just start with the default and add rules as we need them. Not sure if you have any strong opinions here.

**Review by parsonsmatt — DISMISSED** [review]

File: src/Moat/Pretty/Swift.hs [hunk 2]
```diff
@@ -98,13 +95,14 @@ prettyProtocols = \case
         Equatable -> "Equatable"
         OtherProtocol s -> s
 
+-- TODO: Need a plan to avoid @error@ in these pure functions
+{-# ANN prettyTags "HLint: ignore" #-}
```
Comment (chiroptical) [inline src/Moat/Pretty/Swift.hs:99]: This is unfortunate, but I need to form a better plan for the `pretty{Language}` functions to be the same types.

File: src/Moat.hs [hunk 3]
```diff
@@ -144,8 +136,9 @@ type X = Void
 ensureEnabled :: Extension -> ShwiftyM ()
 ensureEnabled ext = do
   enabled <- lift $ isExtEnabled ext
-  unless enabled $ do
-    throwError $ ExtensionNotEnabled ext
+  if enabled
+    then pure ()
+    else throwError $ ExtensionNotEnabled ext
```
Comment (chiroptical) [inline src/Moat.hs:141]: `unless` is super confusing to me. Hilariously, I got the `if` logic backwards the first time.

**Review by parsonsmatt — DISMISSED** [review]

File: src/Moat.hs [hunk 4]
```diff
@@ -1275,19 +1258,20 @@ deriveConstraint ::
...
+    tName :: ShwiftyM Name
     tName = varTToName typ
+    varTToName :: Type -> ShwiftyM Name
     varTToName = \case
-      VarT n -> n
+      VarT n -> pure n
       SigT t _ -> varTToName t
-      _ -> error "Shwifty.varTToName: encountered non-type variable"
-deriveConstraint ClassData _ = Nothing
+      _ -> fail "Shwifty.varTToName: encountered non-type variable"
```
[the remark this answers is not in the document]
Comment (chiroptical) [reply src/Moat.hs:1273]: This doesn't quite work, but I see where you are trying to go! Adding a custom `ShwiftyError` and using `throwError` to make this a bit better.

File: src/Moat.hs [hunk 5]
```diff
@@ -1584,9 +1568,7 @@ tagExp tyconName parentName typ dis =
       (mkName "tagParent", unqualName parentName),
       (mkName "tagTyp", toMoatTypeECxt typ),
       ( mkName "tagDisambiguate",
-        case dis of
-          False -> ConE 'False
-          True -> ConE 'True
+        if dis then ConE 'True else ConE 'False
```
[the remark this answers is not in the document]
Comment (chiroptical) [reply src/Moat.hs:1581]: This gave me a really weird error and then I realized we aren't in `Q` here. You are :100: correct though!

File: src/Moat.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (chiroptical) [reply src/Moat.hs]: handled in d3557ea, but would like to do this for other `fail`s

[the remark this answers is not in the document]
Comment (chiroptical) [reply src/Moat.hs]: the other one is handled here: 7837d23

File: src/Moat.hs (hunk 3 above)
Comment (parsonsmatt) [reply src/Moat.hs:141]: I hate `if` for that reason too! I never get `Bool` stuff right. I really just want custom sum types for everything...

File: src/Moat.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (parsonsmatt) [reply src/Moat.hs:1580]: Oh, right, my bad 😅

**Review by parsonsmatt — APPROVED** [review]

**Pull request merged**